test/extensions/filters/http/adaptive_concurrency/controller/gradient_controller_test.cc

```diff
@@ -786,6 +787,69 @@ TEST_F(GradientControllerTest, ConsecutiveMinConcurrencyReset) {
   }
 }
 
+// Verify interactions in multi-thread latency samples.
+TEST_F(GradientControllerTest, MultiThreadSampleInteractions) {
+  const std::string yaml = R"EOF(
+sample_aggregate_percentile:
+  value: 50
+concurrency_limit_params:
+  max_concurrency_limit:
+  concurrency_update_interval: 0.1s
+min_rtt_calc_params:
+  jitter:
+    value: 0.0
+  interval: 3600s
+  request_count: 5
+  buffer:
+    value: 0
+  min_concurrency: 100
+)EOF";
+
+  auto controller = makeController(yaml);
+  auto& synchronizer = controller->synchronizer();
+  synchronizer.enable();
+
+  for (int i = 0; i < 4; ++i) {
+    tryForward(controller, true);
+    sampleLatency(controller, std::chrono::microseconds(1337));
+  }
+
+  // The next sample will trigger the minRTT value update. We'll spin off a thread and block before
+  // the actual function call to update the value.
+  EXPECT_TRUE(controller->inMinRTTSamplingWindow());
+  synchronizer.waitOn("pre_minrtt_update");
+  synchronizer.waitOn("post_minrtt_update");
+
+  std::thread t1([this, &controller]() {
+    tryForward(controller, true);
+    sampleLatency(controller, std::chrono::microseconds(1337));
+  });
+
+  // Wait for the thread t1 to wait.
+  synchronizer.barrierOn("pre_minrtt_update");
+
+  // We can now kick off another thread to sample another request, but we don't expect this one to
+  // block since thread1 is already in that critical section.
```

Comment:
You need to verify this is the case by joining thread 2 immediately after spawning it, while the pre-minrtt update barrier is in place. If t2 hits the critical section, the test will hang and we have a race condition. Otherwise, it joins and the test continues.

Reply:
Could we chat over slack to discuss more? i am not sure whether I follow.